demos/http/http_demo_basic_tls/http_demo_basic_tls.c

```diff
@@ -0,0 +1,810 @@
+}
+
+/*-----------------------------------------------------------*/
+
+static int tlsSetup( int tcpSocket,
+                     SSL ** pSslContext )
+{
+    int sslStatus = 0;
+    char * cwd = getcwd( NULL, 0 );
+    FILE * pRootCaFile = NULL;
+    X509 * pRootCa = NULL;
+
+    /* Setup for creating a TLS client. */
+    SSL_CTX * pSslSetup = SSL_CTX_new( TLS_client_method() );
+
+    if( pSslSetup != NULL )
+    {
+        /* Set auto retry mode for the blocking calls to SSL_read and SSL_write.
+         * The mask returned by SSL_CTX_set_mode does not need to be checked. */
+        ( void ) SSL_CTX_set_mode( pSslSetup, SSL_MODE_AUTO_RETRY );
+
+        /* OpenSSL does not provide a single function for reading and loading certificates
+         * from files into stores, so the file API must be called. */
+        pRootCaFile = fopen( SERVER_CERT_PATH, "r" );
+
+        /* Check if an absolute directory is being used. */
+        if( ( SERVER_CERT_PATH[ 0 ] == '/' ) || ( SERVER_CERT_PATH[ 0 ] == '\\' ) )
+        {
+            LogInfo( ( "Attempting to open root CA certificate: Path=%.*s.",
+                       ( int32_t ) SERVER_CERT_PATH_LENGTH,
+                       SERVER_CERT_PATH ) );
+        }
+        else
+        {
+            LogInfo( ( "Attempting to open root CA certificate: Path=%s/%.*s.",
+                       cwd,
+                       ( int32_t ) SERVER_CERT_PATH_LENGTH,
+                       SERVER_CERT_PATH ) );
+        }
+
+        if( pRootCaFile != NULL )
+        {
+            pRootCa = PEM_read_X509( pRootCaFile, NULL, NULL, NULL );
+        }
+        else
+        {
+            LogError( ( "Unable to find the root CA certificate file: "
+                        "SERVER_CERT_PATH=%.*s.",
+                        ( int32_t ) SERVER_CERT_PATH_LENGTH,
+                        SERVER_CERT_PATH ) );
+        }
+
+        if( pRootCa != NULL )
+        {
+            /* Add the server's root CA to the set of trusted certificates. */
+            sslStatus = X509_STORE_add_cert( SSL_CTX_get_cert_store( pSslSetup ),
+                                             pRootCa );
+        }
+        else
+        {
+            LogError( ( "Failed to parse the root CA certificate from"
+                        " file %.*s. Please validate the certificate.",
+                        ( int32_t ) SERVER_CERT_PATH_LENGTH,
+                        SERVER_CERT_PATH ) );
+        }
+    }
+
+    /* Set up the TLS connection. */
+    if( sslStatus == 1 )
+    {
+        /* Create a new SSL context. */
+        *pSslContext = SSL_new( pSslSetup );
+
+        if( *pSslContext != NULL )
+        {
+            /* Enable SSL peer verification. */
+            SSL_set_verify( *pSslContext, SSL_VERIFY_PEER, NULL );
+
+            sslStatus = SSL_set_fd( *pSslContext, tcpSocket );
+        }
+        else
+        {
+            LogError( ( "Failed to create a new SSL context." ) );
+            sslStatus = 0;
```

Comment:
All the error log messages will be printed after this point. How about changing the structure like the following:

```c
/* Perform the TLS handshake. */
if( sslStatus == 1 )
{
    sslStatus = SSL_connect( *pSslContext );
    /* Immediately check for error. */
    if( sslStatus != X509_V_OK  ) /* X509_V_OK  or whatever the success code from SSL_connect. */
    {
        LogError( ( "Failed to perform TLS handshake." ) );
    }  
}

if( sslStatus == OK )
{
    /* Perform next operation. */
    /* Log error if failed. */
}
```

This also makes the code more readable by logging the error close to the failure.